File: esp32/components/graphics_lib/graphics/ui/intinput/intinput.cpp
#include "intinput.h"

ui_intinput_config_t UIIntInput::defaultConfig()
{
    return {
        .text = "",
        .customFormat = [](const int value) { return std::to_string(value); },
        .onChange = [](const int value) { return value; },
        .onDone = [](const int _) {}
    };
}

UIIntInput::UIIntInput(ui_intinput_config_t config, const int defaultValue)
    : value(defaultValue), config(std::move(config))
{

}


std::string UIIntInput::render(const bool selected)
{
    return (selected ? ">" : "") +  config.text + ": " + config.customFormat(value);
}

bool UIIntInput::on_event(const graphics_event_t event)
{
    switch (event)
    {
    case EVENT_CLICK:
        focus = !focus;
        old_value = value;
        if (!focus)
            config.onDone(value);
        return focus;
    case EVENT_BACK:
        value = old_value;
        return focus;
    case EVENT_SCROLL_RIGHT:
        value = config.onChange(value + 1);
        return false;
    case EVENT_SCROLL_LEFT:
        value = config.onChange(value - 1);
        return false;
    default:
        return UIElement::on_event(event);
    }
    /*
    if (event == EVENT_FOCUS)
    {
        old_value = value;
        return true;
    }
    if (event == EVENT_SCROLL_RIGHT)
    {
        value++;
        if (value > maxValue)
            value = 200;
        return true;
    }
    if (event == EVENT_SCROLL_LEFT)
    {
        value--;
        if (value < minValue)
            value = 60;
        return true;
    }
    if (event == EVENT_OK)
    {
        onDone(value);
        return false;
    }
    if (event == EVENT_BACK)
    {
        value = old_value;
        return false;
    }
    return UIElement::on_event(event);
    */
}

File: esp32_rust/components/main_rust/src/graphics/ui/intinput.rs
use crate::graphics::event::GraphicsEvent;
use crate::graphics::ui::element::{UIElement, UIElementState};

pub struct IntInputConfig {
    pub text: String,
    pub format_value: Box<dyn Fn(i32) -> String>,
    pub on_change: Box<dyn Fn(i32, i32) -> i32>,
    pub on_done: Box<dyn Fn(i32) -> ()>
}

impl IntInputConfig {
    pub fn new<F1, F2, F3>(
        text: &str,
        format_value: F1,
        on_change: F2,
        on_done: F3
    ) -> Self
    where
        F1: Fn(i32) -> String + 'static,
        F2: Fn(i32, i32) -> i32    + 'static,
        F3: Fn(i32) -> ()     + 'static {
        Self {
            text: String::from(text),
            format_value: Box::new(format_value),
            on_change: Box::new(on_change),
            on_done: Box::new(on_done)
        }
    }
}

impl Default for IntInputConfig {
    fn default() -> Self {
        IntInputConfig {
            text: "Input".to_string(),
            format_value: Box::new(|value| value.to_string()),
            on_change: Box::new(|value, _| value),
            on_done: Box::new(|_| {})
        }
    }
}

pub struct UIIntInput {
    focus: bool,
    config: IntInputConfig,
    old_value: i32,
    value: i32,
}

impl UIIntInput {
    pub fn new(config: IntInputConfig, default_value: i32) -> Self {
        UIIntInput {
            focus: false,
            config,
            old_value: default_value,
            value: default_value
        }
    }
}

impl UIElement for UIIntInput {
    fn render(&self, state: UIElementState) -> String {
        let prefix = match state {
            UIElementState::Selected => if self.focus {
                "-"
            } else {
                ">"
            },
            UIElementState::SelectorPress => "\\",
            UIElementState::None => ""
        };
        format!("{}{}: {}", prefix, self.config.text, (self.config.format_value)(self.value))
    }

    fn on_event(&mut self, event: GraphicsEvent) -> bool {
        match event {
            GraphicsEvent::Click => {
                self.focus = !self.focus;
                if !self.focus && self.value != self.old_value {
                    (self.config.on_done)(self.value);
                }
                self.old_value = self.value;
                self.focus
            }
            GraphicsEvent::Back => {
                self.focus = false;
                self.value = self.old_value;
                true
            }
            GraphicsEvent::ScrollRight => {
                self.value = (self.config.on_change)(self.value + 1, self.value);
                false
            }
            GraphicsEvent::ScrollLeft => {
                self.value = (self.config.on_change)(self.value - 1, self.value);
                false
            }
            _ => false
        }
    }
}